artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/buffer/TimedBuffer.java

```diff
@@ -358,13 +358,7 @@ public boolean flushBatch() {
                   bytesFlushed.addAndGet(pos);
                }
 
-               final ByteBuffer bufferToFlush = bufferObserver.newBuffer(bufferSize, pos);
-               //bufferObserver::newBuffer doesn't necessary return a buffer with limit == pos or limit == bufferSize!!
-               bufferToFlush.limit(pos);
-               //perform memcpy under the hood due to the off heap buffer
-               buffer.getBytes(0, bufferToFlush);
-
-               bufferObserver.flushBuffer(bufferToFlush, pendingSync, callbacks);
+               bufferObserver.flushBuffer(buffer.byteBuf(), pendingSync, callbacks);
```

Comment:
I wish it was that simple.

that's only viable if your write operation is synchronous.
That's probably ok with NIO or Mapped, but it's definitely not ok with libaio.

If you allow the inner buffer to be reused like that, TimedBuffer will be making changes to a buffer that's been submitted to the kernel, and you may endup with wrong data.

And that kind of race is hard to catch.

You can only that do this kind of thing on NIO or Mapped, assuming their buffers will be copied to the file right away.

Reply:
It has been solved on the abstract seq file: ASYNCIO is doing is exactly like before (with the copy), while NIO/MAPPED avoid the copy by using a specialized class

Reply:
Hmmm.. but you moved the copy to LocalBufferObserver::flushBuffer.

What's the difference then, you're still making the copy.

Reply:
I see.. you moved the previous logic to the SequentialFile.. nice one!!

What happens to Mapped though, from what I looked it still using a new buffer. is that what you wanted?

Reply:
By default `AbstractSequentialFile` cannot say if the underline file is performing async or sync writes, hence it just copy it, like it was done in the past on `TimedBuffer`, while NIO/MAPPED can assume that writes are synchronous and they will avoid the copy (if possilble, there are some contraints I've written on the code).

Reply:
> What happens to Mapped
It will use another wrapper I've written some time ago, because `MappedSequentialFile` isn't a child of 'AbstractSequentialFile`, here: https://github.com/apache/activemq-artemis/pull/2522/files#diff-6707e3363162061f97af51ef7f8d13a4R256